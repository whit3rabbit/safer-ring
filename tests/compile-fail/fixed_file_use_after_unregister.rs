//! This test verifies that fixed files cannot be used after being unregistered,
//! ensuring registry safety.

use safer_ring::{Registry, Ring};

#[tokio::main]
async fn main() {
    let mut registry = Registry::new();
    let fixed_files = registry.register_fixed_files(vec![0, 1]).unwrap();
    let fixed_stdin = fixed_files[0].clone();
    
    // Unregister all fixed files
    registry.unregister_fixed_files().unwrap();
    
    // This should fail to compile in a real implementation
    // For now it's a runtime safety issue that we document
    let ring = Ring::new(32).unwrap();
    let buffer = safer_ring::PinnedBuffer::with_capacity(1024);
    
    // Using fixed file after unregistration should be prevented
    let _future = ring.read_fixed(&fixed_stdin, buffer.as_mut_slice()); //~ ERROR: potential use after unregister
}