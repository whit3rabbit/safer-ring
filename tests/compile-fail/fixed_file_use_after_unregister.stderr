error[E0596]: cannot borrow `ring` as mutable, as it is not declared as mutable
  --> tests/compile-fail/fixed_file_use_after_unregister.rs:21:19
   |
21 |     let _future = ring.read_fixed(&fixed_stdin, buffer.as_mut_slice()); //~ ERROR: potential use after unregister
   |                   ^^^^ cannot borrow as mutable
   |
help: consider changing this to be mutable
   |
17 |     let mut ring = Ring::new(32).unwrap();
   |         +++

error[E0596]: cannot borrow `buffer` as mutable, as it is not declared as mutable
  --> tests/compile-fail/fixed_file_use_after_unregister.rs:21:49
   |
21 |     let _future = ring.read_fixed(&fixed_stdin, buffer.as_mut_slice()); //~ ERROR: potential use after unregister
   |                                                 ^^^^^^ cannot borrow as mutable
   |
help: consider changing this to be mutable
   |
18 |     let mut buffer = safer_ring::PinnedBuffer::with_capacity(1024);
   |         +++
